### ISSUE TITLE
CTC Error crashes on empty GPU batch

### ISSUE THREAD

**Issue by michael-s-crouch** [tensorflow/tensorflow#44348]
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04.1 LTS
- TensorFlow installed from (source or binary): binary 
- TensorFlow version (use command below): v2.3.0-54-gfcc4b966f1 2.3.1
- Python version: 3.8.5
- CUDA/cuDNN version: 11.0
- GPU model and memory: GeForce RTX 2080 Ti 11GB x8

When I run the following notebook with 8 GPUs activated, it fails:
https://colab.research.google.com/drive/1yU02l0WtSW2mp7Ai-UvUF18Z2vbNPbzy?usp=sharing

If I run it with 1 GPU, or on CPU, it works. If I change "valnum" from 65 to 64, it works. 

It seems like the CTC Error function doesn't handle empty batches being distributed to GPUs?

The error messages when it fails are mostly variations of:
`Invalid argument:  Tried to stack elements of an empty list with non-fully-defined element_shape: [1,?]
	 [[node replica_6/functional_5/ctc/scan/TensorArrayV2Stack/TensorListStack (defined at <ipython-input-3-6220b8957bcc>:14) ]]
	 [[replica_2/functional_5/ctc/ArithmeticOptimizer/ReorderCastLikeAndValuePreserving_int32_Squeeze_1/_499]]`

Full error log: [errorlog.txt](https://github.com/tensorflow/tensorflow/files/5445445/errorlog.txt)

(This has been a problem on a transfer learning application, where I have a very large training set and need to use all 8 GPUs, but a very small validation set, which crashes if I use all 8 GPUs.)

**Comment by amahendrakar** [comment]
@creidieki,
Each TensorFlow release is compatible with certain CUDA and cuDNN versions. TensorFlow v2.3 is compatible with CUDA 10.1 and cuDNN 7.6, for more information please take a look at the [tested build configurations](https://www.tensorflow.org/install/source#gpu). 

Could you please check if you are facing the same issue with CUDA 10.1 and cuDNN 7.6? Thanks!

**Comment by creidieki** [comment]
To clarify my previous comment, this is happening with the stock Ubuntu packages. I haven't recompiled tensorflow against an unsupported version of libcuda, or force-installed any packages without their dependencies, or anything like that. I really think it's a bug, because it either works or crashes depending entirely on whether one of the GPUs receives an empty batch during training.

**Comment by nikitamaia** [comment]
Hi @creidieki, I was able to reproduce this on GCP with 4 GPUs if I dropped the `valnum` to 33. 32 and 34 worked okay. To try and isolate what could be causing the problem, have you tried this using `tf.data` instead of numpy?

**Comment by creidieki** [comment]
It seems to only happen with the ctc loss. If I change the loss function to:
`def ctc_lambda_func(args):
    return K.constant(1, shape=(24, 1))`
it doesn't crash. 

I don't know about the tf.data question, because the documentation for tf.data.Dataset doesn't cover creating a dataset from a dictionary of named inputs. I tried adding the line 
`inputs = tf.data.Dataset.from_tensor_slices(inputs)`
just before model.predict but it didn't work.

**Comment by creidieki** [comment]
Since this has been confirmed as a bug, could the label "Support" be removed? This is still affecting me. Thanks.

**Comment by Luux** [comment]
We're encountering this issue as well. Did you find any workaround?

**Comment by creidieki** [comment]
I ended up having to rearrange things so that there would never be partial batches. So I've been doing things like:
* choosing a validation set size of 120, so that it would divide evenly among 1, 2, 3, 4, 5, 6, or 8 GPUs
* given n training examples, throwing away the last (n % 32) of them, so that all training batches were size 32
I'm sorry I can't offer anything better.

**Comment by AminHP** [comment]
Any updates on the fix? This is a really annoying bug!

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
